### PR TITLE
Remove legacy unmaintained packages, refactor to support change default locale

### PR DESCRIPTION
**Pull request by wxiaoguang**
Follows 
* https://github.com/go-gitea/gitea/pull/19284#issuecomment-1086137621

This PR removes two unmaintained vendor packages `i18n` and `paginater`

Changes:

* Rewrite `i18n` package with a more clear fallback mechanism. Fix an unstable `Tr` behavior.
* To keep minimal change, the new `i18n` still keeps the old name. In future, many code can be merged together with the `translation` package.
* In future there can be a refactoring that allow the template call `locale.Tr("the.key")` directly, without the redundant `$.Lang`.
* Add more tests for `i18n`.
* Refactor the legacy `Paginater` to `Paginator`, test cases are kept unchanged.

Trivial enhancement (no breaking for end users):

* Use the first locale in LANGS setting option as the default, requested by a user recently (forgot the ID, maybe in discord): `i18n.DefaultLocales.SetDefaultLang(setting.Langs[0])`, and there is a log to prevent from surprising users.

The UI with i18n still works as before (of course ...), screenshot:

<details>

image

</details>

### CODE REVIEW

**Review by lafriks — APPROVED** [review]

**Comment by silverwind** [comment]
One future change regarding `i18n`: It should be made a helper function, not something that's passed with the data, because it is a pain to pass that data down to sub-templates. I imagine a lot of templates could be simplified if it were a helper.

File: modules/paginator/paginator.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+func (p *Paginator) Total() int {
+	return p.total
+}
+
+// TotalPages returns number of total pages.
+func (p *Paginator) TotalPages() int {
+	if p.total == 0 {
+		return 1
+	}
+	return (p.total + p.pagingNum - 1) / p.pagingNum
```
Comment (Gusted) [inline modules/paginator/paginator.go:102]:
```suggestion
	return (p.total - 1)/p.pagingNum + 1
```
Doing the + pagingNum is effectively just adding a one to the end result. So bit useless to do in that way(I know it's copied from the package, but since were having it's own module now...)

File: modules/paginator/paginator.go [hunk 2]
```diff
@@ -0,0 +1,206 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+// Copyright 2015 Unknwon. Licensed under the Apache License, Version 2.0
```
Comment (Gusted) [inline modules/paginator/paginator.go:2]: I'm not a lawyer but I'm not sure if we're allowed to re-license this under MIT-style license(even though we credit it, the source code becomes effectively MIT-licensed source code).

File: modules/translation/i18n/i18n.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+// AddLocale adds locale into the store, and the locale is attached to the store, it can not be added to another one
+func (ls *LocaleStore) AddLocale(lc *locale) error {
+	if _, ok := ls.localeMap[lc.langName]; ok {
+		return ErrLocaleAlreadyExist
+	}
+
+	if lc.store != nil {
+		return ErrLocaleAlreadyAdded
+	}
```
Comment (Gusted) [inline modules/translation/i18n/i18n.go]:
```suggestion
	if _, ok := ls.localeMap[lc.langName]; ok || lc.store != nil {
		return ErrLocaleAlreadyExist
	}
```

File: modules/paginator/paginator.go (hunk 1 above)
Comment (wxiaoguang) [reply modules/paginator/paginator.go:102]: It's not copied from the package, it was `if ...` before, I re-wrote the expression 😂

Why I always use `(total + align - 1) /  align` because the `total` can be unsigned in some cases, `total - 1` then would overflow. And the `total` can never reach around UNSIGNED_INTxx_MAX (it's impossible to hold so much data in real world).

Especially when using C++, all size_t are unsigned, so the only choice is using `(total + align - 1) /  align` there.

In conclusion, I think  `(total + align - 1) /  align` is a very general and safe method to do the math trick, and I hope we can keep it.

File: modules/paginator/paginator.go (hunk 2 above)
Comment (wxiaoguang) [reply modules/paginator/paginator.go:2]: What I mean is before the code was under Apache License. Then newly written code is under MIT license, I did some changes to the code.

To prevent misundestanding, I changed these texts now.

File: modules/translation/i18n/i18n.go (hunk 3 above)
Comment (wxiaoguang) [reply modules/translation/i18n/i18n.go]: ErrLocaleAlreadyExist is not the same as ErrLocaleAlreadyAdded.
ErrLocaleAlreadyExist means the locale name already exists in current store.
ErrLocaleAlreadyAdded means this locale had been added to other store (one locale can not be used in two stores)

And I made a new commit to remove the `AddLocale` from this PR, it isn't necessary at the moment, we can add it back in future if there is an real requirement. Now `-17,+7=-10` lines.

Comment (Gusted) [reply modules/translation/i18n/i18n.go]: *Woops* Didn't notice the different naming.

File: modules/paginator/paginator.go (hunk 1 above)
Comment (Gusted) [reply modules/paginator/paginator.go:102]: > Why I always use `(total + align - 1) / align` because the `total` can be unsigned in some cases, `total - 1` then would overflow. And the `total` can never reach around UNSIGNED_INTxx_MAX (it's impossible to hold so much data in real world).
> 
> Especially when using C++, all size_t are unsigned, so the only choice is using `(total + align - 1) / align` there.
> 
> In conclusion, I think `(total + align - 1) / align` is a very general and safe method to do the math trick, and I hope we can keep it.

But you could overflow if align is some insane high number(close to math.MaxInt64) the suggested change simply avoids that 😄. But it's a edge-case which we don't really have to worry about.

**Review by Gusted — APPROVED** [review]
LGTM! Always great to refactor away packages(less binary size 😉)
